// re-store/dist/esnext/Store/Store.js
import { Map } from 'immutable';
import { createStore as createReduxStore } from 'redux';
function isNode() {
    return Object.prototype.toString.call(typeof process !== 'undefined' ? process : 0) === '[object process]';
}
export default class Store {
    constructor(options) {
        this.reducerObjects = {};
        this.reduce = (previousState, action) => {
            return Object.keys(this.reducerObjects).reduce((nextState, key) => {
                return nextState.updateIn([key], (v) => this.reducerObjects[key].reduce(v, action));
            }, previousState || Map());
        };
        // oop
        this.reduxStore = options && options.redux
            || this.createStore(options && options.preloadState);
        this.isNode = options && options.isNode !== undefined
            ? options.isNode
            : isNode();
        this.isDev = options && options.isDev !== undefined
            ? options.isDev
            : process && process.env && process.env.NODE_ENV === 'development';
        this.useRemoteDevtools = options && !!options.useRemoteDevtools;
    }
    registerReducer(reducerClass) {
        const nReducer = new reducerClass(this);
        if (!this.reducerObjects.hasOwnProperty(nReducer.name)) {
            this.reducerObjects[nReducer.name] = nReducer;
        }
    }
    getState(name) {
        if (name) {
            return this.reduxStore.getState().get(name);
        }
        else {
            return this.reduxStore.getState();
        }
    }
    dispatch(type, payload) {
        this.reduxStore.dispatch({
            payload,
            type,
        });
    }
    createStore(preloadState) {
        try {
            const rde = require('redux-devtools-extension');
            const rrd = require('remote-redux-devtools');
            if (this.isDev && !this.isNode) {
                const devTools = this.useRemoteDevtools ? rrd.devToolsEnhancer : rde.devToolsEnhancer;
                return createReduxStore(this.reduce, preloadState, devTools({}));
            }
            else {
                return createReduxStore(this.reduce, preloadState);
            }
        }
        catch (e) {
            return createReduxStore(this.reduce, preloadState);
        }
    }
}
